### ISSUE TITLE
My opinion

### ISSUE THREAD

**Issue by sagikazarmark** [Happyr/HttpAutoDiscovery#1]
Hey Tobias,

Thanks for reaching out. I've looked through your package. While I see the point, I have some points which might show you why this approach is not so good.

As I can see you already use our static discovery service. While it might not be documented (yet), discovery service itself is a bad idea. It is only created so that you can create a zero config environment if you really need it. It is a fallback. The recommended way is to use dependency injection. That is the only point where knowing what client you are going to use is possible and acceptable.

That being said, the code you've written might serve as an example: how to use the discovery service (for example in an API client). Actually this is how I imagined an example class which use an underlying client: you accept a Client interface argument, but if none is passed, you can find one using the discovery. But the primary target is to use dependency injection.

To sum up: implementing it as an actual client decorator is not so good idea I think, but using a similar code in an API client for example, this is how I imagine it.

One more thing though: why did you reimplement the HTTPMethods trait? It is being moved to a utility repo ATM, but it is still in the master branch.

**Comment by Nyholm** [comment]
Thank you. The point of this library is to use it with API clients. Like this LinkedIn client (https://github.com/Happyr/LinkedIn-API-client/blob/psr7/src/LinkedIn.php#L165). Zero config is needed is this case as you say. I agree with you that discovery should not be used when you are developing an application. 

> One more thing though: why did you reimplement the HTTPMethods trait? It is being moved to a utility repo ATM, but it is still in the master branch.

Great! I would be happy to use that. Are you talking about this repo? https://github.com/php-http/utility

**Comment by sagikazarmark** [comment]
> The point of this library is to use it with API clients.

I think it is better to use an actual client. Adding extra layer to avoid duplicationg code or adding the discovery itself to the API client doesn't worth it IMO. I think the code of this class should directly go into the API client.

See how it is used here: https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/AbstractProxyClient.php#L64

(Still using the adapter naming, will probably change when alpha is out @dbu?)

> Are you talking about this repo? https://github.com/php-http/utility

Yup, but it is still WIP. It already has some content on my computer (migrated from httplug), but haven't pushed it yet. Needs some more tests, etc.
